pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go

```diff
@@ -1013,68 +1149,34 @@ func (t *Test) chooseUpgradePath() ([]*clusterupgrade.Version, error) {
 		return []*clusterupgrade.Version{pred, predPred}, nil
 	}
 
-	// possibleUpgradePathsMap maps a version to the possible upgrade paths that
-	// can be taken with exactly n upgrades.
-	possibleUpgradePathsMap := make(map[string]map[int][][]*clusterupgrade.Version)
-	// findPossibleUpgradePaths finds all legal upgrade paths ending at version `v` with
-	// exactly `numUpgrades` upgrades.
```

Comment:
This whole memoization machinery is no longer needed, so I reverted the original, simpler method. Also, in the case of `{predPred, pred}`, the original method would choose _either_ at random, whereas here you're choosing both, and using random choice at the end; while semantically the same, they induce different prng states; luckily it affected only one unit test.

Reply:
Actually, there was a regression in `findPossibleUpgradePaths`. The original search uses `chosenPredecessor := predecessors[t.prng.Intn(len(predecessors))]`. Thus, the upgrade to the final version is always chosen whereas `findPossibleUpgradePaths` may end up choosing some other path.